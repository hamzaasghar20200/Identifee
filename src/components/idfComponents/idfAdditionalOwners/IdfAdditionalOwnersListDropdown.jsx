import { useEffect, useState } from 'react';
import { Dropdown } from 'react-bootstrap';

import Avatar from '../../Avatar';
import IdfAllAdditionalOwnersList from './IdfAllAdditionalOwnersList';

const IdfAdditionalOwnersListDropdown = ({
  service,
  serviceId,
  maxOwners,
  maxCount = 9,
  refreshOwners,
  defaultSize = 'sm',
  className = '',
  setRefresOwners,
  prevalueCount,
  prevalue,
  preOwners,
  filteredIcons = false,
  ...props
}) => {
  const [openAllOwners, setOpenAllOwners] = useState(false);
  const [owners, setOwners] = useState([]);
  const [count, setCount] = useState(0);
  const [show, setShow] = useState(false);

  useEffect(() => {
    if (prevalue && prevalueCount) {
      setCount(prevalueCount);
      setOwners(preOwners);
    }
  }, [prevalueCount]);

  useEffect(() => {
    if (serviceId) onGetOwners();
  }, [serviceId, refreshOwners]);

  const ownersService = async (pagination) => {
    return await service
      .getOwners(serviceId, pagination)
      .catch((err) => console.log(err));
  };

  const onGetOwners = async () => {
    const resp = await ownersService({ page: 1, limit: 10 });

    const {
      data,
      pagination: { count },
    } = resp || {};

    setOwners(data);
    setCount(count);
  };

  return (
    <Dropdown
      show={show}
      onToggle={(isOpen, event, metadata) => {
        if (metadata.source !== 'select') {
          setShow(isOpen);
        }
      }}
    >
      <Dropdown.Toggle
        className={`btn btn-icon btn-${defaultSize} btn-ghost-primary section-owners-header icon-ignore rounded-circle ${className}`}
      >
        <span
          className={`material-icons-outlined border-0 ${
            defaultSize === 'sm' ? 'fs-6' : ''
          } fs-6`}
        >
          add
        </span>
        <span className={`fw-bold border-0`}>{count - maxOwners}</span>
      </Dropdown.Toggle>

      <Dropdown.Menu className="border border-1">
        {owners?.slice(2, owners.length).map((owner, i) => (
          <div key={owner.user_id}>
            <Dropdown.Item
              data-testid={owner.user_id}
              className="d-flex align-items-center"
            >
              <div className="avatar avatar-sm avatar-circle mr-3">
                <Avatar
                  user={owner.user}
                  style={
                    filteredIcons
                      ? {
                          border: `${
                            props.clicked &&
                            props.clicked.includes(i + maxOwners)
                              ? '2px solid blue'
                              : ''
                          }`,
                          padding: '2px',
                        }
                      : {}
                  }
                />
              </div>
              <h5 className="mb-0">
                {owner.user.first_name
                  ? `${owner.user.first_name} ${owner.user.last_name}`
                  : `${owner.user.email}`}
              </h5>
            </Dropdown.Item>
          </div>
        ))}
        {count > maxCount && (
          <IdfAllAdditionalOwnersList
            openAllOwners={openAllOwners}
            setOpenAllOwners={setOpenAllOwners}
            service={service}
            serviceId={serviceId}
            count={count}
            refreshOwners={refreshOwners}
            setRefresOwners={setRefresOwners}
            {...props}
          >
            {filteredIcons || (
              <Dropdown.Item>
                <div onClick={() => setOpenAllOwners(true)}>
                  View all
                  <span className="material-icons-outlined">chevron_right</span>
                </div>
              </Dropdown.Item>
            )}
          </IdfAllAdditionalOwnersList>
        )}
      </Dropdown.Menu>
    </Dropdown>
  );
};

export default IdfAdditionalOwnersListDropdown;
