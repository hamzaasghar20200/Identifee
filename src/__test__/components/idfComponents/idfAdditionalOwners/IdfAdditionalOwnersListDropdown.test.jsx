import { fireEvent, render, screen } from '@testing-library/react';
import { waitFor } from '@testing-library/dom';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import IdfAdditionalOwnersListDropdown from '../../../../components/idfComponents/idfAdditionalOwners/IdfAdditionalOwnersListDropdown';
import organizationService from '../../../../services/organization.service';
import {
  dataGetOneOwners,
  dataGetThreeOwners,
} from '../../../__mocks__/services/organization.service';

jest.mock('../../../../services/organization.service');

describe('Test IdfAdditionalOwnersListDropdown', () => {
  const props = {
    defaultSize: 'xs',
    service: organizationService,
    maxOwners: 3,
    serviceId: '1',
    mainOwner: {
      id: 1,
      first_name: 'First Name',
      last_name: 'Last Name',
      title: 'Title',
    },
  };
  const history = createMemoryHistory();

  beforeEach(async () => {
    jest.clearAllMocks();
    jest.resetModules();
    jest.restoreAllMocks();
    jest.resetAllMocks();

    await organizationService.getOwners.mockImplementation(() =>
      Promise.resolve(dataGetOneOwners)
    );
  });

  it('Snapshot component without dropdown menu', async () => {
    const component = render(
      <Router history={history}>
        <IdfAdditionalOwnersListDropdown {...props} />
      </Router>
    );

    await waitFor(() => {
      expect(component.container).toBeVisible();
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it('Snapshot component with dropdown menu', async () => {
    const component = render(
      <Router history={history}>
        <IdfAdditionalOwnersListDropdown {...props} />
      </Router>
    );

    const buttonAdd = component.getByText('add');
    fireEvent.click(buttonAdd);

    await screen.findByText(/Owner/i);

    await waitFor(() => {
      expect(component.container).toBeVisible();
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it("Should render button 'View All'", async () => {
    await organizationService.getOwners.mockImplementation(() =>
      Promise.resolve(dataGetThreeOwners)
    );

    const component = render(
      <Router history={history}>
        <IdfAdditionalOwnersListDropdown {...props} maxCount={2} />
      </Router>
    );

    const buttonAdd = component.getByText('add');
    fireEvent.click(buttonAdd);

    await screen.findAllByText(/Owner/i);
    const buttonView = component.getByText('View all');

    await waitFor(() => {
      expect(buttonView).toBeVisible();
      expect(component.baseElement).toMatchSnapshot();
    });
  });
});
