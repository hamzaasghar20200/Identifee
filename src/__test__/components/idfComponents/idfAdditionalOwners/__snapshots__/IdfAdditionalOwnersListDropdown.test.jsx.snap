// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test IdfAdditionalOwnersListDropdown Should render button 'View All' 1`] = `
<body>
  <div>
    <div
      class="show dropdown"
    >
      <button
        aria-expanded="true"
        aria-haspopup="true"
        class="btn btn-icon btn-xs btn-ghost-primary section-owners-header rounded-circle  dropdown-toggle btn btn-primary"
        type="button"
      >
        <span
          class="material-icons-outlined border-0"
        >
          add
        </span>
      </button>
      <div
        aria-labelledby=""
        class="basic-animation border border-1 dropdown-menu show"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, 0px);"
        x-placement="bottom-start"
      >
        <a
          class="d-flex align-items-center dropdown-item"
          data-testid="1"
          href="/settings/profile/users/1"
        >
          <div
            class="avatar avatar-sm avatar-circle mr-3"
          >
            <div
              class="avatar avatar-sm avatar-circle avatar-soft-primary "
            >
              <span
                class="avatar-initials avatar-icon-font-size"
              >
                OA
              </span>
            </div>
          </div>
          <h5
            class="mb-0"
          >
            Owner 1
             
            Additional
          </h5>
        </a>
        <a
          class="d-flex align-items-center dropdown-item"
          data-testid="2"
          href="/settings/profile/users/2"
        >
          <div
            class="avatar avatar-sm avatar-circle mr-3"
          >
            <div
              class="avatar avatar-sm avatar-circle avatar-soft-primary "
            >
              <span
                class="avatar-initials avatar-icon-font-size"
              >
                OA
              </span>
            </div>
          </div>
          <h5
            class="mb-0"
          >
            Owner 2
             
            Additional
          </h5>
        </a>
        <a
          class="d-flex align-items-center dropdown-item"
          data-testid="3"
          href="/settings/profile/users/3"
        >
          <div
            class="avatar avatar-sm avatar-circle mr-3"
          >
            <div
              class="avatar avatar-sm avatar-circle avatar-soft-primary "
            >
              <span
                class="avatar-initials avatar-icon-font-size"
              >
                OA
              </span>
            </div>
          </div>
          <h5
            class="mb-0"
          >
            Owner 3
             
            Additional
          </h5>
        </a>
        <a
          class="dropdown-item"
          href="#"
          role="button"
        >
          <div>
            View all
            <span
              class="material-icons-outlined"
            >
              chevron_right
            </span>
          </div>
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`Test IdfAdditionalOwnersListDropdown Snapshot component with dropdown menu 1`] = `
<body>
  <div>
    <div
      class="show dropdown"
    >
      <button
        aria-expanded="true"
        aria-haspopup="true"
        class="btn btn-icon btn-xs btn-ghost-primary section-owners-header rounded-circle  dropdown-toggle btn btn-primary"
        type="button"
      >
        <span
          class="material-icons-outlined border-0"
        >
          add
        </span>
      </button>
      <div
        aria-labelledby=""
        class="basic-animation border border-1 dropdown-menu show"
        data-popper-escaped="true"
        data-popper-placement="bottom-start"
        data-popper-reference-hidden="true"
        style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, 0px);"
        x-placement="bottom-start"
      >
        <a
          class="d-flex align-items-center dropdown-item"
          data-testid="1"
          href="/settings/profile/users/1"
        >
          <div
            class="avatar avatar-sm avatar-circle mr-3"
          >
            <div
              class="avatar avatar-sm avatar-circle avatar-soft-primary "
            >
              <span
                class="avatar-initials avatar-icon-font-size"
              >
                OA
              </span>
            </div>
          </div>
          <h5
            class="mb-0"
          >
            Owner 1
             
            Additional
          </h5>
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`Test IdfAdditionalOwnersListDropdown Snapshot component without dropdown menu 1`] = `
<body>
  <div>
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="btn btn-icon btn-xs btn-ghost-primary section-owners-header rounded-circle  dropdown-toggle btn btn-primary"
        type="button"
      >
        <span
          class="material-icons-outlined border-0"
        >
          add
        </span>
      </button>
    </div>
  </div>
</body>
`;
